plugin-server/src/main/ingestion-queues/session-recording/services/session-manager-v3.ts

```diff
@@ -0,0 +1,472 @@
+                await readFile(path.join(context.dir, 'metadata.json'), 'utf-8')
+            )
+            manager.buffer = {
+                context: bufferMetadata,
+                fileStream: manager.createFileStreamFor(path.join(context.dir, BUFFER_FILE_NAME)),
+            }
+        } catch (error) {
+            // Indicates no buffer metadata file or it's corrupted
+            status.error('🧨', '[session-manager] failed to read buffer metadata', {
+                ...context,
+                error,
+            })
+        }
+
+        status.info('📦', '[session-manager] started new manager', {
+            ...manager.context,
+            ...(manager.buffer?.context ?? {}),
+        })
+        return manager
+    }
+
+    private async save(): Promise<void> {
+        if (this.buffer) {
+            await writeFile(this.file('metadata.json'), JSON.stringify(this.buffer?.context), 'utf-8')
+        } else {
+            await unlink(this.file('metadata.json'))
+        }
+    }
+
+    private async getFlushFiles(): Promise<string[]> {
+        return (await readdir(this.context.dir)).filter((file) => file.endsWith(FLUSH_FILE_EXTENSION))
+    }
+
+    private captureException(error: Error, extra: Record<string, any> = {}): void {
+        captureException(error, {
+            extra: { ...this.context, ...extra },
+            tags: { teamId: this.context.teamId, sessionId: this.context.sessionId },
+        })
+    }
+
+    private captureMessage(message: string, extra: Record<string, any> = {}): void {
+        const context = this.context
+        captureMessage(message, {
+            extra: { ...context, ...extra },
+            tags: { teamId: context.teamId, sessionId: context.sessionId },
+        })
+    }
+
+    public async add(message: IncomingRecordingMessage): Promise<void> {
+        if (this.destroying) {
+            return
+        }
+
+        try {
+            const buffer = this.getOrCreateBuffer()
+            const messageData = convertToPersistedMessage(message)
+            const start = message.events.at(0)?.timestamp
+            const end = message.events.at(-1)?.timestamp ?? start
+
+            if (!start || !end) {
+                captureMessage("[session-manager]: can't set events range from message without events summary", {
+                    extra: { message },
+                })
+                return
+            }
+
+            buffer.context.eventsRange = {
+                firstTimestamp: minDefined(start, buffer.context.eventsRange?.firstTimestamp) ?? start,
+                lastTimestamp: maxDefined(end, buffer.context.eventsRange?.lastTimestamp) ?? end,
+            }
+
+            const content = JSON.stringify(messageData) + '\n'
+            buffer.context.count += 1
+            buffer.context.sizeEstimate += content.length
+
+            await this.save()
+
+            if (!buffer.fileStream.write(content, 'utf-8')) {
```

Comment:
so we're updating the buffer with count/length/timetamps before trying to write the file
if the buffer write succeeds and then filter write fails then the buffer is incorrect?
need to store original buffer content so you can revert that change in exception handler?

Reply:
Yeah there is a race here and its unclear which is best.

Reply:
feels like the most significant change...

there's an implicit transaction in "update metadata, save content, commit to kakfa". anything that goes wrong in there is the bit where we'll risk getting unplayable data

it's very tricky because `try { writeA(new); writeB(new) } catch { writeA(original); /* no need to revert B because content can have duplicates */ }` if write new fails it may be that we can't write the original because this consumer has lost its access to disk or network or something

can the metadata be in the file or in file headers or something? so there's one write to disk that did or didn't succeed?